lib/jekyll/cache.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require "digest"
+require "fileutils"
+require "pstore"
+
+module Jekyll
+  class Cache
+    extend Forwardable
+
+    # Get an existing named cache, or create a new one if none exists
+    #
+    # name - name of the cache
+    #
+    # Returns nothing.
+    # rubocop:disable Style/ClassVars
+    def initialize(name)
+      @@base_dir ||= File.expand_path(".jekyll-cache/Jekyll/Cache")
+      @@caches ||= {}
+      @@safe ||= true
+      @cache = @@caches[name] ||= {}
+      @name = name
+      FileUtils.mkdir_p(path_to) if @@safe
+    end
+
+    def self.clear
+      delete_cache_files
+      @@caches.each_value(&:clear)
+    end
+
+    def self.safe(safe = true)
+      @@safe = safe unless defined?(@safe) && safe.nil?
+    end
+    # rubocop:enable Style/ClassVars
+
+    def clear
+      delete_cache_files
+      @cache.clear
+    end
+
+    def [](key)
+      return @cache[key] if @cache.key?(key)
+      path = path_to(hash(key))
+      if @@safe && File.file?(path) && File.readable?(path)
+        @cache[key] = load(path)
+      else
+        raise
+      end
+    end
+
+    def getset(key)
+      return @cache[key] if @cache.key?(key)
+      path = path_to(hash(key))
+      if @@safe && File.file?(path) && File.readable?(path)
+        value = load(path)
```

Comment:
I really wanted to just call `#[]` from `#getset`, but it won't work. They both need to do very different things in the else block.

Reply:
Agreed.

Reply:
_Or_, since `#[]` just throws if it the key doesn't exist, `#getset` could `rescue` and do it's thing? I don't know if this would make the code easier or more difficult to follow.

```ruby
def getset(key)
  self[key]
rescue
  value = yield
  dump(path, value) if @@safe
  @cache[key] = value
end
```